pkg/crc/machine/sync.go

```diff
@@ -0,0 +1,119 @@
+package machine
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/code-ready/crc/pkg/crc/logging"
+	"github.com/code-ready/machine/libmachine/state"
+	"golang.org/x/sync/semaphore"
+)
+
+const startCancelTimeout = 15 * time.Second
+
+type Synchronized struct {
+	startLock      *semaphore.Weighted
+	stopDeleteLock *semaphore.Weighted
+	startCancel    context.CancelFunc
+	underlying     Client
+}
+
+func NewSynchronizedMachine(machine Client) *Synchronized {
+	return &Synchronized{
+		startLock:      semaphore.NewWeighted(1),
+		stopDeleteLock: semaphore.NewWeighted(1),
+		underlying:     machine,
+	}
+}
+
+func (s *Synchronized) IsProgressing() bool {
+	if !s.startLock.TryAcquire(1) {
+		return true
+	}
+	defer s.startLock.Release(1)
+	if !s.stopDeleteLock.TryAcquire(1) {
+		return true
+	}
+	defer s.stopDeleteLock.Release(1)
+	return false
+}
+
+func (s *Synchronized) Delete() error {
+	cleanup, err := s.lockForStopOrDelete()
+	defer cleanup()
+	if err != nil {
+		return err
+	}
+	return s.underlying.Delete()
+}
+
+func (s *Synchronized) Start(ctx context.Context, startConfig StartConfig) (*StartResult, error) {
+	if !s.startLock.TryAcquire(1) {
+		return nil, errors.New("cluster is starting")
+	}
+	defer s.startLock.Release(1)
+	ctx, s.startCancel = context.WithCancel(ctx)
+	return s.underlying.Start(ctx, startConfig)
+}
+
+func (s *Synchronized) Stop() (state.State, error) {
+	cleanup, err := s.lockForStopOrDelete()
+	defer cleanup()
+	if err != nil {
+		return state.Error, err
+	}
+	return s.underlying.Stop()
+}
+
+func (s *Synchronized) lockForStopOrDelete() (func(), error) {
+	if !s.stopDeleteLock.TryAcquire(1) {
+		return func() {
+
+		}, errors.New("cluster is stopping or deleting")
+	}
+
+	if s.startCancel != nil {
+		logging.Infof("Cancelling virtual machine start...")
+		s.startCancel()
+	}
+
```

Comment:
Fwiw, I think this has a slight race, if the `Start()` operation completes at about the same time as when we call `s.startCancel()`, another `Start()` might be started right away before this method acquires the `startLock` a few lines below. Probably not going to matter much though.

Reply:
Yes, I don't see how to fix it correctly. Perhaps run `startLock.Acquire` first and in parallel run `cancelStart`.

Reply:
I think its' correct in https://github.com/cfergeau/crc/blob/sync/pkg/crc/machine/sync.go :)

Reply:
Ok I took your commit. I converted the enum to string as it is easier to read/debug.